src/indexer/index_sorter.rs

```diff
@@ -0,0 +1,426 @@
+use super::SegmentWriter;
+use crate::{DocId, IndexSettings, Order, TantivyError};
+use std::cmp::Reverse;
+
+/// Struct to provide mapping from old docid to new docid and vice versa
+pub struct DocidMapping {
+    new_docid_to_old: Vec<u32>,
+    old_docid_to_new: Vec<u32>,
+}
+
+impl DocidMapping {
+    /// returns the new docid for the old docid
+    pub fn get_new_docid(&self, docid: DocId) -> DocId {
+        self.old_docid_to_new[docid as usize]
+    }
+    /// returns the old docid for the new docid
+    pub fn get_old_docid(&self, docid: DocId) -> DocId {
+        self.new_docid_to_old[docid as usize]
+    }
+    /// iterate over old docids in order of the new docids
+    pub fn iter_old_docids(&self) -> std::slice::Iter<'_, u32> {
+        self.new_docid_to_old.iter()
+    }
+}
+
+pub(crate) fn sort_index(
+    settings: IndexSettings,
+    segment_writer: &mut SegmentWriter,
+) -> crate::Result<DocidMapping> {
+    let docid_mapping = get_docid_mapping(settings, segment_writer)?;
+    Ok(docid_mapping)
+}
+
+// Generates a document mapping in the form of [index new docid] -> old docid
+fn get_docid_mapping(
+    settings: IndexSettings,
+    segment_writer: &mut SegmentWriter,
```

Comment:
is mut required here?

Reply:
`get_field_writer` required &mut. I made some changes, it's non mut now